grug/types/src/length_bounded.rs

```diff
@@ -0,0 +1,371 @@
+        }
+
+        if length > MAX {
+            return Err(StdError::length_out_of_bound::<T>(length, ">", MAX));
+        }
+
+        Ok(Self(value))
+    }
+
+    pub fn new_unchecked(value: T) -> Self {
+        Self(value)
+    }
+}
+
+impl<T, const MIN: usize, const MAX: usize> Inner for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds,
+{
+    type U = T;
+
+    fn inner(&self) -> &Self::U {
+        &self.0
+    }
+
+    fn into_inner(self) -> Self::U {
+        self.0
+    }
+}
+
+impl<T, const MIN: usize, const MAX: usize> AsRef<T> for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds,
+{
+    fn as_ref(&self) -> &T {
+        &self.0
+    }
+}
+
+impl<T, const MIN: usize, const MAX: usize> Deref for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds,
+{
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T, const MIN: usize, const MAX: usize> Serialize for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds + Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+impl<'de, T, const MIN: usize, const MAX: usize> Deserialize<'de> for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds + Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value = T::deserialize(deserializer)?;
+
+        LengthBounded::new(value).map_err(serde::de::Error::custom)
+    }
+}
+
+impl<T, const MIN: usize, const MAX: usize> BorshSerialize for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds + BorshSerialize,
+{
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        self.0.serialize(writer)
+    }
+}
+
+impl<T, const MIN: usize, const MAX: usize> BorshDeserialize for LengthBounded<T, MIN, MAX>
+where
+    T: LengthBounds + BorshDeserialize,
+{
+    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
+        let value = BorshDeserialize::deserialize_reader(reader)?;
+
+        Self::new(value).map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+    }
+}
+
+// ---------------------------- LengthBounds impls -----------------------------
+
+impl LengthBounds for Binary {
+    fn length(&self) -> usize {
+        self.len()
+    }
+}
+
+impl LengthBounds for String {
+    fn length(&self) -> usize {
+        self.len()
+    }
+}
+
+impl LengthBounds for Vec<u8> {
+    fn length(&self) -> usize {
+        self.len()
+    }
+}
+
+impl LengthBounds for &[u8] {
+    fn length(&self) -> usize {
+        self.len()
+    }
+}
+
+impl<const LEN: usize> LengthBounds for [u8; LEN] {
+    fn length(&self) -> usize {
+        self.len()
+    }
+}
+
+impl LengthBounds for &str {
+    fn length(&self) -> usize {
+        self.len()
+    }
+}
```

Comment:
No need to impl for `&[u8]` and `&str` because we never use them in json messages (they don't implement `serde::Deserialize`).

For `[u8; LEN]`, can you change it to `[T; LEN]`